scripts/o3de/o3de/compatibility.py

```diff
@@ -0,0 +1,295 @@
+    """
+    if (object_json_data.get('compatible_engines') or object_json_data.get('engine_api_dependencies')):
+        return engine_is_compatible(
+            engine_json_data['engine_name'], engine_json_data.get('version',''), 
+            object_json_data.get('compatible_engines',''), object_json_data.get('engine_api_dependencies',''),
+            engine_json_data)
+    return True
+
+def engine_is_compatible(engine_name:str, engine_version:str, compatible_engines:list, engine_api_version_specifiers:list, engine_json_data:dict or None) -> bool:
+    """
+    Returns True if the engine is compatible with the provided compatible_engines and engine_api_versions information.
+    If a compatible_engine entry only has an engine name, it is assumed compatible with every engine version with that name.
+    If an engine_api_version entry only has an api name, it is assumed compatible with every engine api version with that name.
+    :param engine_name: the engine name
+    :param engine_version: the engine version
+    :param compatible_engines: a list of engine names and (optional)version specifiers
+    :param engine_api_versions: a list of engine api names and (optional)version specifiers
+    :param engine_json_data: optional engine json data to avoid extra file ops
+    """
+    # early out if there are no restrictions
+    if not compatible_engines and not engine_api_version_specifiers:
+        return True
+
+    if not engine_json_data:
+        if engine_name:
+            engine_json_data = manifest.get_engine_json_data(engine_name=engine_name)
+        else:
+            engine_json_data = manifest.get_engine_json_data(engine_path=manifest.get_this_engine_path())
+
+    if not engine_json_data:
+        return False
+
+    if not engine_name:
+        # fallback to use the information from the current engine
+        engine_name = engine_json_data.get('engine_name','')
+        engine_version = engine_json_data.get('version','')
+
+    if compatible_engines:
+        if engine_version and has_compatible_version(compatible_engines, engine_name, engine_version):
+            return True
+        elif not engine_version and has_compatible_name(compatible_engines, engine_name):
+            # assume an engine with no version is compatible
+            return True
+
+    engine_api_versions = engine_json_data.get('api_versions','')
+    if engine_api_version_specifiers and not engine_api_versions:
+        # assume not compatible if no engine api version information is available
+        return False
+
+    if engine_api_version_specifiers:
+        for api_version_specifier in engine_api_version_specifiers:
+            api_name, unused_version_specifiers = utils.get_object_name_and_optional_version_specifier(api_version_specifier)
+            if has_compatible_version([api_version_specifier], api_name, engine_api_versions.get(api_name,'')):
+                return True
+
+    return False
```

Comment:
The if compatible_engines block shouldn't `return True` here as the engine api version might not match.

What we should do is store a boolean that indicates that the engine compatibility check has succeeded so far.
```suggestion
    is_compatible = not compatible_engines
    if compatible_engines:
        if engine_version and has_compatible_version(compatible_engines, engine_name, engine_version):
            is_compatible = True
        elif not engine_version and has_compatible_name(compatible_engines, engine_name):
            # assume an engine with no version is compatible
            is_compatible = True

    if is_compatible:
        engine_api_versions = engine_json_data.get('api_versions','')
        if engine_api_version_specifiers and not engine_api_versions:
            # assume not compatible if no engine api version information is available
            return False

        if engine_api_version_specifiers:
            for api_version_specifier in engine_api_version_specifiers:
                api_name, unused_version_specifiers = utils.get_object_name_and_optional_version_specifier(api_version_specifier)
                if has_compatible_version([api_version_specifier], api_name, engine_api_versions.get(api_name,'')):
                    return True

    return False
```

Reply:
My thinking here was that if an engineer lists some engine in the `compatible_engines` list then it should be considered compatible even if there is some `api` incompatibility - so the fields could be used as an OR instead of AND.  An engine is compatible if it satisfies the `compatible_engines` OR `engine_api_dependencies` OR both.